std/typecons.d

```diff
@@ -4865,12 +4865,154 @@ unittest
     assert(rc1 == 5);
     // No more allocation, add just one extra reference count
     auto rc2 = rc1;
+    assert(rc1.refCountedStore.refCount == 2);
     // Reference semantics
     rc2 = 42;
     assert(rc1 == 42);
     // the pair will be freed when rc1 and rc2 go out of scope
 }
 
+// recursively destroy a class instance
+private void _destroy(T)(T obj) if (is(T == class))
+{
+     foreach (C; TypeTuple!(T, BaseClassesTuple!T))
+     {
+         static if (__traits(hasMember, C, "__dtor"))
+             (cast(C)obj).__dtor();
+     }
+     // leak monitor b/c destruction would be impure, will be collected by the GC
+}
+
+/// RefCounted implementation for classes.
+struct RefCounted(T)
+    if (is(T == class))
+{
+    /// $(D RefCounted) storage implementation.
+    struct RefCountedStore
+    {
+        struct Store
+        {
+            size_t _count;
+            typeof(&._destroy!T) _destroy; // polymorph destructor
+            ubyte[__traits(classInstanceSize, T)] _payload;
+        }
+        Store* _store;
+
+        /// Returns `true` if the underlying store has been allocated and initialized.
+        @property nothrow @safe
+        bool isInitialized() const
+        {
+            return _store !is null;
+        }
+
+        /// Get the underlying reference count.
+        @property nothrow @safe
+        size_t refCount() const
+        {
+            return _store._count;
+        }
+
+    private:
+
+        void destroy()
+        {
+            import core.memory : GC;
+            import core.stdc.stdlib : free;
+
+            _store._destroy(cast(T)_store._payload.ptr);
+            static if (hasIndirections!T)
+                GC.removeRange(_store._payload.ptr);
+            free(_store);
+            _store = null;
+        }
+    }
+    private RefCountedStore _refCounted;
+
+    /// Returns storage implementation struct.
+    @property nothrow @safe
+    ref inout(RefCountedStore) refCountedStore() inout
+    {
+        return _refCounted;
+    }
+
+    /// Boolean conversion for RefCounted, true when initialized.
+    bool opCast(T:bool)() const @safe
```

Comment:
This notation catches things like `cast(int)` as well, it's probably better to use `bool opCast(T)() if(is(T == bool))`